### PR TITLE
Do not announce crafting progress loss when none happened

### PR DESCRIPTION
**Pull request by pehamm**
#### Summary
Bugfixes "Do not show crafting progress loss when none happened"

#### Purpose of change

During crafting, the game sets failure points at which failures happens. This is done via a routine that determines among other things whether progress was lost. However, while the code allows for the possibility that no progress is lost at all, it will always print a message announcing a craft fail and progress loss, even if none happened at all ("You mess up and lose 0% progress.")

Steps to reproduce:

1. Create a debug character with 2 tailoring skill, 1 survival skill no proficiency in "Principles of Leatherworking"
2. Start and immediately stop 5 leather funnels.
3. Debug failoring and survival skill to 10 and add "Principles of Leatherworking" proficiency.
4. Pick up the in-progress crafts and finish them.
5. Almost all of the objects should display the "You mess up and lose 0% progress." message exactly once, never with any actual progress loss.

#### Describe the solution

Check whether there is actual progress loss. If there is none, skip the message (and the unneccessary adjustment to the currenct status by subtracting 0).

#### Describe alternatives you've considered

None. Simple bug with an obvious fix.

#### Testing

Tested the above replication steps and received no notifications over 0% progress loss. Also verified that normal progress loss still happens, only messages when there is no actual loss are no longer shown.

#### Additional context

Note that the message showing 0% progress lost can still appear due to rounding, but only if the actual amount is less than 0.5% but larger than 0%.

### CODE REVIEW

File: src/crafting.cpp [hunk 1]
```diff
@@ -1096,10 +1096,12 @@ bool item::handle_craft_failure( Character &crafter )
...
                                          ( 1.0 - std::min( success_roll, 1.0 ) );
-    const int progess_loss = item_counter * percent_progress_loss;
-    crafter.add_msg_player_or_npc( _( "You mess up and lose %d%% progress." ),
-                                   _( "<npcname> messes up and loses %d%% progress." ), progess_loss / 100000 );
-    item_counter = clamp( item_counter - progess_loss, 0, 10000000 );
+    if( percent_progress_loss > 0.0 ) {
+        const int progess_loss = item_counter * percent_progress_loss;
+        crafter.add_msg_player_or_npc( _( "You mess up and lose %d%% progress." ),
+                                       _( "<npcname> messes up and loses %d%% progress." ), progess_loss / 100000 );
+        item_counter = clamp( item_counter - progess_loss, 0, 10000000 );
```
Comment (Night-Pryanik) [inline src/crafting.cpp:1103]:
```suggestion
        const int progress_loss = item_counter * percent_progress_loss;
        crafter.add_msg_player_or_npc( _( "You mess up and lose %d%% progress." ),
                                       _( "<npcname> messes up and loses %d%% progress." ), progress_loss / 100000 );
        item_counter = clamp( item_counter - progress_loss , 0, 10000000 );
```

Comment (pehamm) [reply src/crafting.cpp]: Fixed, thanks!

**Review by Night-Pryanik — APPROVED** [review]

File: src/crafting.cpp [hunk 2]
```diff
@@ -1096,10 +1096,12 @@ bool item::handle_craft_failure( Character &crafter )
     // Loss is scaled by the success roll
     const double percent_progress_loss = rng_exponential( 0.25, 0.35 ) *
                                          ( 1.0 - std::min( success_roll, 1.0 ) );
```
Comment (actual-nh) [inline src/crafting.cpp:1098]:
```suggestion
    const double percent_progress_loss = rng_exponential( 0.25, 0.35 ) * ( 1.0 - success_roll );
```
Since it isn't used if below 0.

Comment (pehamm) [reply src/crafting.cpp:1098]: You are right mathematically, good catch. I am a bit wary because it loses the property of being a loss in that case, but maybe a comment up front is enough to clarify that. The std::min was actually added as a bugfix two years ago to prevent negative loss (or random progress advances during "failures"), but my solution fixes that as well as you noted. Just have to ensure that future commiters are warned up front about this.

Comment (pehamm) [reply src/crafting.cpp]: Done, plus added a new comment and edited an existing one for greater clarity.

**Review by actual-nh — APPROVED** [review]
